elit0451/SI-FinalProject
# Notification service - Issue #38 #39 #40 #41

## Checklist
- [x] Implement a RESTful endpoint to fetch new notifications
- [x] Implement a RESTful endpoint to mark notification(s) as read
- [x] Subscribe to meaningful services (creation service)
- [x] Connect to a SQL Database

### PR Description
Created a new .Net Core Web API.
Created a DBSchema for the service, which is described in "DBSCHEMA" file located in the root.
Added Notification model which corresponds to the Notification entity in the DB.
Added connection to MySQL within the project.
Added connection to RabbitMQ within the project.
Added a Get endpoint to fetch new notifications.
Added a Put endpoint to mark notifications as read.
Added functionality subscribe to "event.add" queue.
Added Dockerfile to build an image from Notification service.

### How to test
I have created a docker image of the notification service and it is pushed to docker hub

Start by executing the following docker commands:
- `docker network create testnet`
- `docker run --rm --name mysql -d -v ${PWD}/db_data:/var/lib/mysql -p 3306:3306 --network=testnet -e MYSQL_ROOT_PASSWORD=soft2019Backend mysql`
**Make sure to execute the DBSCHEMA file against the database to have the proper data.**

- `docker run --rm --name rabbitmq -d --network=testnet -p 15672:15672 -p 5672:5672 rabbitmq:management`
- `docker run --rm --name notificationservice -p 5002:5002 --network=testnet davi7816/testnotificationservice`
(before succeeding to start the service, you should give some time to rabbitmq to start)

To add a new event, and get a notification for it, head to http://localhost:15672/, queues. Go into **event.add** and send a similar message:
```
{"EventId":5,
"EventType":{"Name":"Sally's birthday"},
"DateFrom":"2020-02-23",
"DateTo":"2020-02-23",
"Location":"Santorini"}
```

To update an event, and get a notification for it, head to http://localhost:15672/, queues. Go into **event.update** and send a similar message:
```
{"EventId":5,
"cmd":"car",
"license":"1234567890"}
```
and check the DB for those 2 new notifications.

To test the functionality for marking a notification as read, head to Postman (or similar tool) and execute a PUT request by `http://localhost:5002/Notification/{id}`. Make sure the `markedRead` attribute is set to true.

To test the functionality for getting notifications for a specific event, do a GET request to `http://localhost:5002/Notification/{eventId}`

Issue reference:

close #38  
close #39  
close #40  
close #41
